### PR TITLE
SOLR-16006: Fix declared/unused dependency warnings

### PR DESCRIPTION
**Pull request by risdenk**
https://issues.apache.org/jira/browse/SOLR-16006

Most of the changes here go from `*implementation` to `*runtimeOnly` or add an exception. This cleans up most of the warnings in all the projects except for the `extraction` module (see SOLR-15989) and the `gcs-repository` module (see SOLR-15990).

### CODE REVIEW

File: solr/modules/s3-repository/build.gradle [hunk 1]
```diff
@@ -20,70 +20,70 @@ apply plugin: 'java-library'
 description = 'S3 Repository'
 
 dependencies {
-    api project(':solr:core')
-    implementation project(':solr:solrj')
+  api project(':solr:core')
+  implementation project(':solr:solrj')
 
-    implementation 'org.apache.lucene:lucene-core'
+  implementation 'org.apache.lucene:lucene-core'
```
Comment (dsmiley) [inline solr/modules/s3-repository/build.gradle:26]: if we make this "api" on solr-core, then would we _still_ need to declare this here?

File: solr/server/build.gradle [hunk 2]
```diff
@@ -30,39 +30,62 @@ configurations {
...
+  permitUnusedDeclared 'io.dropwizard.metrics:metrics-graphite'
   api('io.dropwizard.metrics:metrics-graphite', {
     exclude group: "com.rabbitmq", module: "amqp-client"
   })
+  permitUnusedDeclared 'io.dropwizard.metrics:metrics-jetty9'
   api 'io.dropwizard.metrics:metrics-jetty9'
+  permitUnusedDeclared 'io.dropwizard.metrics:metrics-jvm'
   api 'io.dropwizard.metrics:metrics-jvm'
+  permitUnusedDeclared 'io.dropwizard.metrics:metrics-jmx'
   api 'io.dropwizard.metrics:metrics-jmx'
```
Comment (dsmiley) [inline solr/server/build.gradle:46]: Maybe these shouldn't be `api` in the first place?

**Review by dsmiley — APPROVED** [review]

File: solr/server/build.gradle (hunk 2 above)
Comment (risdenk) [reply solr/server/build.gradle:46]: So to be completely honest with you - I don't quite understand `solr/server/build.gradle` and how it is different than like core with all the Jetty dependencies. I don't even know if the `api` here even makes any difference. @janhoy on SOLR-16007 mentioned `server/lib` and `WEB-INF/lib` and I don't quite understand the difference between the two right now. So I don't really feel super comfortable with changing from `api` to anything else until I can understand why this was configured this way. (there is some weird webapp -> core -> server dependency stuff going on)

File: solr/modules/s3-repository/build.gradle (hunk 1 above)
Comment (risdenk) [reply solr/modules/s3-repository/build.gradle:26]: That is 100% a valid question and one I don't know the answer to. I think there are a few other implementation -> api that we could make at `core`, `test-framework` and maybe a few others that would make the downstream consumers dependency declaration a bit cleaner. I could try it out, but didn't want to make too many unnecessary changes here.

**Pull request merged**